### DISCUSSION TITLE
Is there a way to specify the DNS used by http_client?

### DISCUSSION THREAD

**Discussion by helkaluin** [jcorporation/myMPD#460] [Q&A] [answered]
Since 7.0 dropped, local playback is now restricted to talking to a port. In a Docker setup where both myMPD and MPD are containerised (in separate containers and connected to the same bridge network) myMPD is able to talk to MPD by resolving "mpd", the same as what you set as comtainer_name for the MPD container. Docker embeds itself as the DNS server within the network, resolving container names so containers can talk to each other.

http_client, however, complains that it cannot find a DNS server, and defaults to Google's:

```
mympd    | WARN     mympd     /myMPD/src/http_client.c:31: No dns servers found
mympd    | WARN     mympd     /myMPD/src/web_server.c:77: Error reading dns server settings
mympd    | DEBUG    mympd     /myMPD/src/web_server.c:79: Setting dns server to udp://8.8.8.8:53
```

which results in Mongoose failing to setup local playback:

```
mympd    | INFO     webserver /myMPD/src/web_server.c:312: HTTP request (20): /stream/
mympd    | INFO     webserver /myMPD/src/web_server.c:326: Creating new mpd stream proxy backend connection to http://mpd:8000
mympd    | INFO     webserver /myMPD/src/web_server.c:340: Forwarding client connection 20 to backend connection 21
mympd    | 2021-03-30 10:46:31  E mongoose.c:411:mg_error   21  DNS lookup failed
mympd    | INFO     webserver /myMPD/src/web_server.c:261: Backend HTTP connection 21 closed
mympd    | INFO     webserver /myMPD/src/web_server.c:476: HTTP connection 20 closed
```

This is odd, since myMPD itself successfully connects to MPD by resolving "mpd" --- mpd_host is set as "mpd". (Cannot use socket since when containerised, MPD fails at binding to a socket.)

There is a DNS server defined in /etc/resolv.conf:

```
# docker exec mympd cat /etc/resolv.conf
nameserver 127.0.0.11
options edns0 trust-ad ndots:0
```

So it seems http_client somehow cannot see this...? As a temporary fix, is it possible for the user to define the DNS used by http_client?

**Comment by jcorporation** [comment, marked as answer]
There is no way to configure a dns server for myMPD.

myMPD uses `res_init()` to get the nameserver from resolv.conf. This is working for gnu libc, but not for musl. The docker container is based on Alpine Linux.

In musl, the `res_init()`function looks like: http://git.musl-libc.org/cgit/musl/tree/src/network/res_init.c. Therefore myMPD finds no dns server. Mongoose as http client library uses his own dns client implementation to resolve names asynchronously and depends on the configured dns server.

Libmpdclient uses `getaddrinfo` to resolve host names and not depends on a working `res_init()` function.

Problem will be fixed for the 7.0.1 version: https://github.com/jcorporation/myMPD/issues/461
